scrapy/selector/unified.py

```diff
@@ -63,7 +68,7 @@ class Selector(_ParselSelector, object_ref):
     __slots__ = ["response"]
     selectorlist_cls = SelectorList
 
-    def __init__(self, response=None, text=None, type=None, root=None, **kwargs):
+    def __init__(self, response=None, text=None, type=None, root=_NOT_SET, **kwargs):
```

Comment:
I would try to avoid importing Parsel’s `_NOT_SET`.

For example, we could declare our own _NOT_SET, and handle root though the kwargs variable, i.e. if root is not _our_ _NOT_SET, we add it to kwargs:

```python
_NOT_SET = object()
...
def __init__(self, response=None, text=None, type=None, root=_NOT_SET, **kwargs):
...
if root is not _NOT_SET:
    kwargs["root"] = root
...
super().__init__(text=text, type=st, **kwargs)
```

Not a strong opinion, though. If one project can use a private member of Parsel that’s Scrapy.

Reply:
On current parsel version(1.8.1) I definitely don't see option without import of `_NOT_SET` from parsel module. Any other internally created `object()` on root - will not pass `_NOT_SET` checks on parsel side

I tried similar approach
https://github.com/scrapy/scrapy/compare/master...GeorgeA92:scrapy:warning_roottext_from_parser_jmespath

Reply:
The key is in _not_ passing `root` at all if set to _NOT_SET on Scrapy’s end.

To elaborate on the code I shared above:

```python
_NOT_SET = object()

...

def __init__(self, response=None, text=None, type=None, root=_NOT_SET, **kwargs):
        if response is not None and text is not None:
            raise ValueError(
                f"{self.__class__.__name__}.__init__() received "
                "both response and text"
            )
        st = _st(response, type)
        if text is not None:
            response = _response_from_text(text, st)
        if response is not None:
            text = response.text
            kwargs.setdefault("base_url", response.url)
        self.response = response
        if root is not _NOT_SET:
            kwargs["root"] = root
        super().__init__(text=text, type=st, **kwargs)
```

Reply:
Very clever, thanks @Gallaecio 😮

Reply:
@Gallaecio @Laerte 
Technically it will work.
Hovewer I don't like any option that make this check on each call of Selector `__init__`.

On spider like this:
```python

    def start_requests(self):
        yield scrapy.Request(url='http://quotes.toscrape.com/', callback=self.parse)

    def parse(self, response):
        links = response.css('a::attr(href)')
        if next_page := response.css("li.next a::attr(href)").get():
            yield scrapy.Request(url=response.urljoin(next_page), callback=self.parse)
```

`__init__` of selector will be called ~50 times per response. On a real case it can be hundreds or even thousands calls of `Selector.__init__` per response.
On applying this - application will make additional +hudnreds/thousands checks of `_NOT_SET` per response. While on initial proposal - related code check will be executed only one time on startup (+root assignment on init). However i din't make detailed profiler tests yet.